nuxeo-features/nuxeo-platform-oauth/src/main/java/org/nuxeo/ecm/platform/oauth2/providers/FrameIOOAuth2ServiceProvider.java

```diff
@@ -0,0 +1,167 @@
+
+import java.io.IOException;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.text.CharacterPredicates;
+import org.apache.commons.text.RandomStringGenerator;
+import org.nuxeo.ecm.core.api.LockHelper;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.kv.KeyValueService;
+import org.nuxeo.runtime.kv.KeyValueStore;
+
+import com.google.api.client.auth.oauth2.AuthorizationCodeFlow;
+import com.google.api.client.auth.oauth2.BearerToken;
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.BasicAuthentication;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpExecuteInterceptor;
+
+/**
+ * @since 11.0
+ */
+public class FrameIOOAuth2ServiceProvider extends NuxeoOAuth2ServiceProvider {
+
+    public static final String FRAMEIO_KEY_VALUE_STORE_NAME = "frameio";
+
+    public static final String FRAMEIO_OAUTH2_KVS_KEY = "nuxeo.provider.frameio.oauth2.key";
+
+    public static final String FRAMEIO_OAUTH2_KVS_STATE_KEY = "nuxeo.provider.frameio.oauth2.state";
+
+    private static final RandomStringGenerator GENERATOR = new RandomStringGenerator.Builder().filteredBy(
+            CharacterPredicates.LETTERS, CharacterPredicates.DIGITS).withinRange('0', 'z').build();
+
+    @Override
+    public String getAuthorizationUrl(HttpServletRequest request) {
+        return getAuthorizationCodeFlow().newAuthorizationUrl()
+                                         .setState(getState())
+                                         .setRedirectUri(getCallbackUrl(request))
+                                         .build();
+    }
+
+    @Override
+    public String getAuthorizationUrl(String serverURL) {
+        return getAuthorizationCodeFlow().newAuthorizationUrl()
+                                         .setState(getState())
+                                         .setRedirectUri(getCallbackUrl(serverURL))
+                                         .build();
+    }
+
+    @Override
+    public Credential handleAuthorizationCallback(HttpServletRequest request) {
+
+        // Checking if there was an error such as the user denied access
+        String error = getError(request);
+        if (error != null) {
+            throw new NuxeoException("There was an error: \"" + error + "\".");
+        }
+
+        // Checking conditions on the "code" URL parameter
+        String code = getAuthorizationCode(request);
+        if (code == null) {
+            // FIXME: improve the error log: error_hint, error_description
+            throw new NuxeoException("There is not code provided as QueryParam.");
+        }
+
+        String state = StringUtils.isBlank(request.getParameter(STATE_PARAM)) //
+                ? null
+                : request.getParameter(STATE_PARAM);
+        checkState(state);
```

Comment:
Sure we can do it, to avoid copy past. But if we do it, it will be only adding the method and the check not the state generation, at least as a first step :)